Training/interface/DataLoader_main.h

```diff
@@ -0,0 +1,770 @@
+           x_grid[CellObjectType::Muon][1] = std::vector<float>(n_tau * n_inner_cells * n_inner_cells * muon_fn,0);
+         }
+
+    std::vector<float> x_tau;
+    std::map<CellObjectType, std::map<bool, std::vector<float>>> x_grid; // [enum class CellObjectType][ 0 - outer, 1 - inner]
+    std::vector<float> weight;
+    std::vector<int> y_onehot;
+};
+
+
+class DataLoader {
+
+public:
+    using Tau = tau_tuple::Tau;
+    using TauTuple = tau_tuple::TauTuple;
+
+    DataLoader() :
+        n_tau(setup::n_tau), current_entry(setup::start_dataset), start_dataset(setup::start_dataset), end_dataset(setup::end_dataset),
+        n_inner_cells(setup::n_inner_cells), inner_cell_size(setup::inner_cell_size), n_outer_cells(setup::n_outer_cells),
+        outer_cell_size(setup::outer_cell_size), n_threads(setup::n_threads), parity(setup::parity),
+        n_fe_tau(setup::n_fe_tau), n_pf_el(setup::n_pf_el) ,n_pf_mu(setup::n_pf_mu),
+        n_pf_chHad(setup::n_pf_chHad), n_pf_nHad(setup::n_pf_nHad) ,n_pf_gamma(setup::n_pf_gamma),
+        n_electrons(setup::n_ele), n_muons(setup::n_muon), n_labels(setup::tau_types),
+        // trainingWeightFactor(tauTuple.GetEntries() / training_weight_factor,
+        innerCellGridRef(n_inner_cells, n_inner_cells, inner_cell_size, inner_cell_size),
+        outerCellGridRef(n_outer_cells, n_outer_cells, outer_cell_size, outer_cell_size)
+    {
+      if(n_threads > 1) ROOT::EnableImplicitMT(n_threads);
+    }
+
+    void Initialize(std::string file_name)
+    {
+        if(tauTuple)
+            throw std::runtime_error("DataLoader is already initialized.");
+        file = OpenRootFile(file_name);
+        tauTuple = std::make_shared<tau_tuple::TauTuple>(file.get(), true);
+    }
+
+    bool HasNext() {
+        return (current_entry + n_tau) < end_dataset;
+    }
+
+    std::shared_ptr<Data> LoadNext(){
+
+        if(!tauTuple)
+            throw std::runtime_error("DataLoader is not initialized.");
+
+        auto data = std::make_shared<Data>(n_tau, n_fe_tau,
+                                           n_inner_cells, n_outer_cells,
+                                           n_pf_el,
+                                           n_pf_mu,
+                                           n_pf_chHad,
+                                           n_pf_nHad,
+                                           n_pf_gamma,
+                                           n_electrons,
+                                           n_muons,
+                                           n_labels
+                                          );
+
+        const Long64_t end_entry = current_entry + n_tau;
+        size_t n_processed = 0, n_total = static_cast<size_t>(end_entry - start_dataset);
+        for(Long64_t tau_i = 0; tau_i < n_tau; ++tau_i, ++current_entry) {
+            tauTuple->GetEntry(current_entry);
+            const auto& tau = tauTuple->data();
+            if(parity == -1 || tau.evt % 2 == static_cast<unsigned>(parity)) {
```

Comment:
We have training/testing split at the level of shuffle&merge. Here it is redundant. Also, the current implementation is wrong because it would not load n_tau in Data if parity != -1.